kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/slot-updates.k

```diff
@@ -0,0 +1,149 @@
+
+    // 1. Slot masking using &Int
+    rule [mask-b-and]:
+      MASK:Int &Int SLOT:Int =>
+        #asWord ( #buf ( 32, SLOT ) [ 32 -Int ( #getMaskShiftBytes(MASK) +Int #getMaskWidthBytes(MASK) ) := #buf ( #getMaskWidthBytes(MASK), 0 ) ] )
+        requires #isMask(MASK) andBool #rangeUInt(256, SLOT)
+        [simplification, concrete(MASK), preserves-definedness]
+
+    // 2a. |Int and +Bytes, update to be done in left
+    rule [bor-update-to-left]:
+      A |Int #asWord ( B1 +Bytes B2 ) =>
+        #asWord ( #buf ( 32 -Int lengthBytes(B2), (A /Int (2 ^Int (8 *Int lengthBytes(B2)))) |Int #asWord ( B1 ) ) +Bytes B2 )
+        requires #rangeUInt(256, A) andBool A modInt (2 ^Int (8 *Int lengthBytes(B2))) ==Int 0 andBool lengthBytes(B1 +Bytes B2) <=Int 32
+        [simplification, preserves-definedness]
+
+    // 2b. |Int of +Bytes, update to be done in right
+    rule [bor-update-to-right]:
+      A |Int #asWord ( B1 +Bytes B2 ) =>
+        #asWord ( B1 +Bytes #buf ( lengthBytes(B2), A |Int #asWord ( B2 ) ) )
+        requires 0 <=Int A andBool A <Int 2 ^Int (8 *Int lengthBytes(B2)) andBool lengthBytes(B2) <=Int 32
+        [simplification, preserves-definedness]
+
+    // 3a. Update with explicit shift and symbolic slot
+    rule [bor-update-with-shift]:
+      ( SHIFT *Int X ) |Int Y => #asWord ( #buf( 32 -Int ( log2Int(SHIFT) /Int 8 ), X ) +Bytes #buf( log2Int(SHIFT) /Int 8, Y ) )
+      requires #isByteShift(SHIFT)
+       andBool 0 <=Int X andBool X <Int 2 ^Int (8 *Int (32 -Int ( log2Int(SHIFT) /Int 8 )))
+       andBool 0 <=Int Y andBool Y <Int SHIFT
+       [simplification, concrete(SHIFT), comm, preserves-definedness]
+
+    // 3b. Buffer cropping
+    rule [buf-asWord-crop]:
+      #buf (W1:Int, #asWord(#buf(W2, X:Int) +Bytes B)) => #buf(W1 -Int lengthBytes(B), X) +Bytes B
+      requires 0 <=Int W1 andBool W1 <=Int 32 andBool 0 <=Int W2 andBool
+       andBool lengthBytes(B) <=Int W1 andBool W1 <=Int W2 +Int lengthBytes(B)
+       andBool 0 <=Int X andBool X <Int 2 ^Int (8 *Int (W1 -Int lengthBytes(B)))
+      [simplification]
+
+    // 3c. Splitting the updated buffer into the updated value and the trailing zeros, explicit shift
+    rule [buf-split-l]:
+      #buf ( W, SHIFT *Int X ) => #buf( W -Int ( log2Int(SHIFT) /Int 8 ), X ) +Bytes #buf( log2Int(SHIFT) /Int 8, 0)
+      requires 0 <=Int W andBool W <=Int 32 andBool #isByteShift(SHIFT)
+       andBool 0 <=Int X andBool X <Int 2 ^Int (8 *Int (W -Int ( log2Int(SHIFT) /Int 8)))
+       [simplification, concrete(W, SHIFT), preserves-definedness]
+
+    // 3d. Splitting the updated buffer into the updated value and the trailing zeros, implicit shift
+    rule [bor-split]:
+      X |Int Y => #asWord ( #buf ( 32 -Int #getFirstOneBit(X) /Int 8, X /Int ( 2 ^Int ( 8 *Int ( #getFirstOneBit(X) /Int 8 ) ) ) ) +Bytes
+                            #buf ( #getFirstOneBit(X) /Int 8, Y ) )
+      requires #rangeUInt(256, X) andBool 0 <=Int #getFirstOneBit(X)
```

Comment:
I guess that `#rangeUInt(256, X)` and `0 <=Int #getFirstOneBit(X)` are somewhat redundant, given that if `notBool #rangeUInt(256, X)`, then #getFirstOneBit(X) => -1`.

So it could be either `#rangeUInt(256, X) andBool notBool X ==Int 0`, or it could be just `0 <=Int #getFirstOneBit(X)`. If you want to keep the original expression for readability that seems fine too.

Reply:
Also, I think a potential counterexample to this one is:

- `X == 2`
- `Y == 1`
- `X |Int Y == 3`
- `#buf(32 -Int #getFirstOneBit(X) /Int 8, X /Int (2 ^Int (8 *Int (#getFirstOneBit(X) /Int 8)))) => #buf(32 -Int 0, X /Int (2 ^Int (8 *Int 0)))) => #buf(32, X /Int 2) => #buf(32, 1)`
- `#asWord(#buf(32, 1)) => 1`


So we have `3 == 1`.

I think restricting `#getFirstOneBit(X) modInt 8 ==Int 0` would suffice, which that condition alone would replace both the first two conditions.

Reply:
Yes, the counter-example holds, will think.

Reply:
```
    rule [bor-split]:
      X |Int Y => #asWord ( #buf ( 32 -Int #getFirstOneBit(X) /Int 8, X /Int ( 2 ^Int ( 8 *Int ( #getFirstOneBit(X) /Int 8 ) ) ) ) +Bytes
                            #buf ( #getFirstOneBit(X) /Int 8, Y ) )
      requires #rangeUInt(256, X) andBool 0 <=Int #getFirstOneBit(X)
       andBool 0 <=Int Y andBool Y <Int 2 ^Int ( 8 *Int ( #getFirstOneBit(X) /Int 8 ) )
```
I don't think that the counter-example works. The last clause
```
(H): Y <Int 2 ^Int ( 8 *Int ( #getFirstOneBit(X) /Int 8 ) )
```
should say that `Y` stops at most in the last byte before `X` begins. In the counter-example, we have that:
- `X ==Int 2`
- `#getFirstOneBit(X) ==Int 2`
- `#getFirstOneBit(X) /Int 8 ==Int 0`
- `2 ^Int ( 8 *Int ( #getFirstOneBit(X) /Int 8 ) ) == 2 ^Int 0 ==Int 1`
and so (H) becomes `Y <Int 1`, which is in contradiction with `Y ==Int 1` of the example.